### PR TITLE
Test service

### PR DESCRIPTION
**Pull request by aslakknutsen**
- [x] Read git-sha to avoid deploying ':latest' v of ike-workspace-test when generating scenarios
- [x] Update e2e test to use new test-service response
- [x] Update e2e test to replace the test-service code

### CODE REVIEW

File: cmd/test-service/handler.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// CallStack holds the complete transaction stack
+type CallStack struct {
+	Caller    string      `json:"caller"`
+	Path      string      `json:"path"`
+	StartTime time.Time   `json:"startTime"`
+	EndTime   time.Time   `json:"endTime"`
+	Called    []CallStack `json:"called,omitempty"`
+}
+
+// NewBasic constructs a new basic HandlerFunc that behaves as described0 by the provided Config
```
Comment (bartoszmajsak) [inline cmd/test-service/handler.go]:
```suggestion
// NewBasic constructs a new basic HandlerFunc that behaves as described by the provided Config
```

File: cmd/test-service/main.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	// EnvHTTPAddr name of Env variable that sets the listning address
```
Comment (bartoszmajsak) [inline cmd/test-service/main.go]:
```suggestion
	// EnvHTTPAddr name of Env variable that sets the listening address
```

File: cmd/test-service/main.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	EnvHTTPAddr = "HTTP_ADDR"
+
+	// EnvServiceName name of Env variable that sets the Service name used in call stack
+	EnvServiceName = "SERVICE_NAME"
+
+	// EnvServiceCall name of Env variable that holds a colon separated list of URLs to call
```
Comment (bartoszmajsak) [inline cmd/test-service/main.go]:
```suggestion
	// EnvServiceCall name of Env variable that holds a colon-separated list of URLs to call
```

File: e2e/infra/istio_setup.go [hunk 4]
```diff
@@ -18,17 +18,30 @@ func LoadIstio() {
...
 	projectDir := os.Getenv("CUR_DIR")
+	registry = setDockerEnvForTestServiceBuild(namespace)
+
+	<-cmd.Execute("oc login -u admin -p admin").Done()
+	<-cmd.ExecuteInDir(".", "bash", "-c", "docker login -u $(oc whoami) -p $(oc whoami -t) "+registry).Done()
+	<-cmd.ExecuteInDir(projectDir, "make", "docker-build-test", "docker-push-test").Done()
+	return
+}
+
+// DeployTestScenario1 deploys book info sample application into specified namespace
```
Comment (bartoszmajsak) [inline e2e/infra/istio_setup.go]:
```suggestion
// DeployTestScenario1 deploys book-info like sample application into the specified namespace
```

File: e2e/infra/istio_setup.go [hunk 5]
```diff
@@ -41,27 +54,69 @@ func DeployOperator() (namespace string) {
...
 
 	<-cmd.ExecuteInDir(projectDir, "make", "deploy-operator").Done()
 	return
 }
 
-func setDockerEnv() (operatorNS, dockerRegistry string) {
-	operatorNS = "istio-workspace-operator"
-	dockerRegistry = "docker-registry-default.127.0.0.1.nip.io:80"
+func setDockerEnvForTestServiceBuild(namespace string) (registry string) {
+	setExampleNamespace(namespace)
```
Comment (bartoszmajsak) [inline e2e/infra/istio_setup.go]: Maybe `setTestNamespace` ?

File: e2e/infra/istio_setup.go [hunk 6]
```diff
@@ -41,27 +54,69 @@ func DeployOperator() (namespace string) {
...
+}
+
+func setDockerEnvForOperatorDeploy() (namespace, registry string) {
+	ns := setOperatorNamespace()
+	repo := setDockerRegistryInternal()
+	return ns, repo
+}
+
+func setExampleNamespace(namespace string) {
+	err := os.Setenv("EXAMPLE_NAMESPACE", namespace)
```
Comment (bartoszmajsak) [inline e2e/infra/istio_setup.go]: Rename would imply changing the env var too.

File: cmd/test-scenario/main.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+var (
+	testImageName = ""
+)
+
+func main() {
+	if len(os.Args) <= 1 {
+		fmt.Println("required arg 'scenario name' missing")
+		os.Exit(-100)
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply cmd/test-scenario/main.go]: Nope, none what so ever. Just < 0 :)

File: cmd/test-scenario/scenarios.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package main
+
+// TestScenario1 is a basic test setup with a few services calling each other in a chain. Similar to the original bookinfo example setup.
+func TestScenario1() {
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply cmd/test-scenario/scenarios.go]: sure, that sounds good.

File: e2e/infra/istio_setup.go (hunk 6 above)
Comment (aslakknutsen) [reply e2e/infra/istio_setup.go]: https://github.com/Maistra/istio-workspace/pull/134/commits/d7b05a328cea2439ff5e98e1c7a9fdb760e3ccc5

File: e2e/infra/istio_setup.go (hunk 5 above)
Comment (aslakknutsen) [reply e2e/infra/istio_setup.go]: https://github.com/Maistra/istio-workspace/pull/134/commits/d7b05a328cea2439ff5e98e1c7a9fdb760e3ccc5

**Comment by aslakknutsen** [comment]
> I was wondering about one thing though. We have more and more references to env vars (i.e. IKE_DOCKER_REGISTRY) - maybe it would be worth to make one place where we keep them as consts?

Agree https://github.com/Maistra/istio-workspace/issues/142

**Pull request merged**